staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go

```diff
@@ -0,0 +1,1175 @@
+						map[string]interface{}{"k": "b", "v": '2'},
+					},
+					[]interface{}{
+						map[string]interface{}{"k": "b", "v": '2'},
+						map[string]interface{}{"k": "a", "v": '1'},
+					},
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"objs": listType(listMapTypePtr([]string{"k"}, objectTypePtr(map[string]schema.Structural{
+					"k": stringType,
+					"v": stringType,
+				}))),
+			}),
+			valid: []string{
+				"objs[0] == objs[1]", // equal even though order is different
+			},
+			errors: map[string]string{
+				"objs[0] == {'k': 'a', 'v': '1'}": "no matching overload for '_==_'", // objects cannot be compared against a data literal map
+			},
+		},
+		{name: "maps",
+			obj:    objs(map[string]interface{}{"k1": "a", "k2": 'b'}, map[string]interface{}{"k2": 'b', "k1": "a"}),
+			schema: schemas(mapType(&stringType), mapType(&stringType)),
+			valid: []string{
+				"val1 == val2", // equal even though order is different
+				"'k1' in val1",
+				"!('k3' in val1)",
+			},
+		},
+		{name: "objects",
+			obj: map[string]interface{}{
+				"objs": []interface{}{
+					map[string]interface{}{"f1": "a", "f2": "b"},
+					map[string]interface{}{"f1": "a", "f2": "b"},
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"objs": listType(objectTypePtr(map[string]schema.Structural{
+					"f1": stringType,
+					"f2": stringType,
+				})),
+			}),
+			valid: []string{
+				"objs[0] == objs[1]",
+			},
+			errors: map[string]string{
+				"objs[0] == {'f1': 'a', 'f2': 'b'}": "no matching overload for '_==_'", // objects cannot be compared against a data literal map
+			},
+		},
+		{name: "object access",
+			obj: map[string]interface{}{
+				"a": map[string]interface{}{
+					"b": 1,
+					"d": 2,
+				},
+				"a1": map[string]interface{}{
+					"b1": map[string]interface{}{
+						"c1": 4,
+					},
+				},
+				"a3": map[string]interface{}{},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"a": *objectTypePtr(map[string]schema.Structural{
+					"b": integerType,
+					"c": integerType,
+					"d": withNullable(true, integerType),
+					"e": withDefault("x", stringType),
+				}),
+				"a1": *objectTypePtr(map[string]schema.Structural{
+					"b1": *objectTypePtr(map[string]schema.Structural{
+						"c1": integerType,
+					}),
+					"d2": *objectTypePtr(map[string]schema.Structural{
+						"e2": integerType,
+					}),
+				}),
+				"a3": *objectTypePtr(map[string]schema.Structural{
+					"b3": withDefault(map[string]interface{}{"c3": "y"}, *objectTypePtr(map[string]schema.Structural{
+						"c3": stringType,
+					})),
+				}),
+			}),
+			// https://github.com/google/cel-spec/blob/master/doc/langdef.md#field-selection
+			valid: []string{
+				"has(a.b)",
+				"!has(a.c)",
+				"has(a.d)",
+				"has(a.e)",
+				"has(a1.b1.c1)",
+				"!(has(a1.d2) && has(a1.d2.e2))", // must check intermediate optional fields (see below no such key error for d2)
+				"!has(a1.d2)",
+			},
```

Comment:
we should also test that a3 doesn't have a default b3

Reply:
I've added:

```
has(a3.b3)
has(a3.b3.c3)
```

Now that the test suite calls defaulting before validation (I didn't always have that), they both pass.